### ISSUE TITLE
vim.system zombie process 

### ISSUE THREAD

**Issue by Ramilito** [neovim/neovim#29475]
### Problem

I'm experiencing zombie process when exiting neovim even though detach is set to false.
```pstree -p``` when neovim is still open: 

![image](https://github.com/neovim/neovim/assets/8473233/d2d52ecd-3878-4c3a-be2a-b6757a007a80)

```pstree -p``` when neovim is closed:

![image](https://github.com/neovim/neovim/assets/8473233/99351765-ac01-435b-94b2-7be3486e8269)

### Steps to reproduce

nvim --clean
:lua vim.system({ "kubectl", "proxy", "--port=0" }, {detach = false})
:q!

### Expected behavior

I'm expecting the child process to be terminated when the parent is terminated if detached is set to false.
If I'm understanding the docs about detached correctly, I need to set it to true to make it live beyond the parents lifetime:
```
• detach: (boolean) If true, spawn the child process in a
detached state - this will make it a process group
leader, and will effectively enable the child to keep
running after the parent exits. Note that the child
process will still keep the parent's event loop alive
unless the parent process calls |uv.unref()| on the
child's process handle.
```

I have a workaround to manually kill it on VimLeavePre but I doubt that will run if neovim crashes:
```lua
  vim.api.nvim_create_autocmd("VimLeavePre", {
    callback = M.stop_kubectl_proxy(),
  })
```

### Neovim version (nvim -v)

v0.11.0-dev

### Vim (not Nvim) behaves the same?

I don't know

### Operating system/version

Windows 11 (WSL)

### Terminal name/version

Wezterm/20240203-110809-5046fc22

### $TERM environment variable

xterm-256color

### Installation

sudo apt-get install neovim

**Comment by lewis6991** [comment]
I'm not sure `detach` has the same semantics on windows. This very well could be a luv bug.

**Comment by Ramilito** [comment]
> I'm not sure detach has the same semantics on windows. I'm not sure how WSL plays into this. One would think it should behave as Linux does. 

Yeah, just verified in OSX and it also leaves the process alive after neovim shutsdown.

> Even with the correct behaviour of detach, there's little we can do to ensure all processes are killed in the event of a crash.

I understand, would be nice to handle it in the normal shutdown though.

**Comment by lewis6991** [comment]
> Yeah, just verified in OSX and it also leaves the process alive after neovim shutsdown.

That's surprising. My understanding was the `detach=false` would always kill all immediate child jobs of Nvim when it is closed (even in a crash). This is supposed to be handled by the OS.

Can you test with a simpler process? `{ "kubectl", "proxy", "--port=0" }` may be launching its own child processes, which Nvim cannot manage. If that's the case, then you may actually need `deatch=true`.

**Comment by gpanders** [comment]
> Yeah, just verified in OSX and it also leaves the process alive after neovim shutsdown.

The OS will not automatically kill an orphaned process. It is reparented to the init process which waits for the orphaned process to exit. If a child process is connected to a pipe from the parent process (as is usually the case for an LSP server), then the child process will exit when its parent dies because the pipe is closed. But if this isn’t the case then the child will happily run forever until it’s told to stop.

**Comment by justinmk** [comment]
> I am not sure what magic libuv does behind the curtain though.

Nothing magic. Nvim kills its immediate children on exit: https://github.com/neovim/neovim/blob/862679c70fa062e567fd7f720f9fd994fa023101/src/nvim/event/process.c#L131-L142

> My understanding was the `detach=false` would always kill all immediate child jobs of Nvim when it is closed (even in a crash). This is supposed to be handled by the OS.

On unix we always need to set `UV_PROCESS_DETACHED`, see explanation in https://github.com/neovim/neovim/commit/8d90171f8be6b92d7186ca84c42a0e07c2c71908

https://github.com/neovim/neovim/blob/862679c70fa062e567fd7f720f9fd994fa023101/src/nvim/event/libuv_process.c#L39-L40

And as a result, 

- on exit we need to actively kill immediate children (then the OS will kill the child's "tree"), if the `vim.system` "job" was started with `detach=false`
    - *internally* we always set `detach=true` for unixes so that libuv does `setsid`. We only *simulate* `detach=false` on exit (and yes this means crashes leave zombies; unavoidable).
- for `detach=true` we simply [skip](https://github.com/neovim/neovim/blob/862679c70fa062e567fd7f720f9fd994fa023101/src/nvim/event/process.c#L136-L138) the above.


Related: https://github.com/neovim/neovim/issues/26861
